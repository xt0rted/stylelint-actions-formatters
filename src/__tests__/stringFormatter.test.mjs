/**
 * Based on https://github.com/stylelint/stylelint/blob/16.0.2/lib/formatters/__tests__/stringFormatter.test.mjs
 */
import process from 'node:process';

import { jest } from '@jest/globals';

import { getCleanFormatterOutput } from './getCleanOutput.mjs';
import stringFormatter from '../stringFormatter.mjs';

describe('stringFormatter', () => {
  let actualTTY;
  let actualColumns;

  beforeAll(() => {
    jest.spyOn(process, 'cwd').mockReturnValue('/test-project');

    actualTTY = process.stdout.isTTY;
    actualColumns = process.stdout.columns;

    // This is needed when running on github actions
    delete process.env.GITHUB_WORKSPACE;
  });

  afterAll(() => {
    process.stdout.isTTY = actualTTY;
    process.stdout.columns = actualColumns;
  });

  describe('using cwd', () => {
    runTestSuite();
  });

  describe('using GITHUB_WORKSPACE', () => {
    beforeAll(() => {
      process.env.GITHUB_WORKSPACE = '/test-project/path';
    });

    afterAll(() => {
      delete process.env.GITHUB_WORKSPACE;
    });

    runTestSuite();
  });

  function runTestSuite() {
    it('outputs no warnings', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [],
        },
      ];

      const output = stringFormatter(results);

      expect(output).toBe('');
    });

    it('outputs warnings', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs warnings contains non-ASCII characters', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar',
              severity: 'error',
              text: '简体中文こんにちは안녕하세요',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('removes rule name from warning text', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'rule-name',
              severity: 'warning',
              text: 'Unexpected foo (rule-name)',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs warnings without stdout `TTY`', () => {
      process.stdout.isTTY = false;

      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar',
              severity: 'error',
              text: 'Unexpected foo',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs warnings with more than 80 characters and `process.stdout.columns` equal 90 characters', () => {
      // For Windows tests
      process.stdout.isTTY = true;
      process.stdout.columns = 90;

      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar-very-very-very-very-very-long',
              severity: 'error',
              text: 'Unexpected very very very very very very very very very very very very very long foo',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs warnings with more than 80 non-ASCII characters and `process.stdout.columns` equal 90 characters', () => {
      // For Windows tests
      process.stdout.isTTY = true;
      process.stdout.columns = 90;

      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar-very-very-very-very-very-long',
              severity: 'error',
              text:
                '简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文' +
                'こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요简体中文こんにちは안녕하세요',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('condenses deprecations and invalid option warnings', () => {
      const results = [
        {
          source: 'file.css',
          deprecations: [
            {
              text: 'Deprecated foo.',
              reference: 'bar',
            },
          ],
          invalidOptionWarnings: [
            {
              text: 'Unexpected option for baz',
            },
          ],
          warnings: [],
        },
        {
          source: 'file2.css',
          deprecations: [
            {
              text: 'Deprecated foo.',
              reference: 'bar',
            },
          ],
          invalidOptionWarnings: [
            {
              text: 'Unexpected option for baz',
            },
          ],
          warnings: [],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('handles ignored file', () => {
      const results = [
        {
          source: 'file.css',
          warnings: [],
          ignored: true,
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toBe('');
    });

    it('handles empty messages', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          warnings: [
            {
              line: 1,
              column: 1,
              rule: 'bar',
              severity: 'error',
              text: '',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });

    it('outputs parse errors and warnings without rule and severity', () => {
      const results = [
        {
          source: '/test-project/path/to/file.css',
          parseErrors: [
            {
              line: 1,
              column: 1,
              stylelintType: 'foo-error',
              text: 'Cannot parse foo',
            },
          ],
          warnings: [
            {
              line: 3,
              column: 1,
              rule: 'no-bar',
              severity: 'error',
              text: 'Disallow bar',
            },
            {
              line: 2,
              column: 1,
              text: 'Anonymous error',
            },
          ],
        },
      ];

      const output = getCleanFormatterOutput(results, stringFormatter);

      expect(output).toMatchSnapshot();
    });
  }
});
