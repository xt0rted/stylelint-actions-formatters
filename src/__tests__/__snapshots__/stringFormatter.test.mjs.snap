// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringFormatter using GITHUB_WORKSPACE condenses deprecations and invalid option warnings 1`] = `
"Invalid Option: Unexpected option for baz

Deprecation warnings:
 - Deprecated foo. See: bar"
`;

exports[`stringFormatter using GITHUB_WORKSPACE handles empty messages 1`] = `
"to/file.css
  1:1  ×     bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs fixable error and warning counts 1`] = `
"to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ×  Unexpected bar  no-bar
  1:2  ‼  Unexpected baz  no-baz

× 3 problems (2 errors, 1 warning)
  1 error and 1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs fixable error counts 1`] = `
"to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ×  Unexpected bar  no-bar

× 2 problems (2 errors, 0 warnings)
  1 error potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs fixable warning counts 1`] = `
"to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)
  1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs fixable warning counts with invalid or missing ruleMetadata 1`] = `
"to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar
  1:2  ‼  Unexpected baz  no-baz

× 3 problems (1 error, 2 warnings)
  1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs parse errors and warnings without rule and severity 1`] = `
"to/file.css
  1:1  ×  Cannot parse foo  foo-error
  2:1  ×  Anonymous error
  3:1  ×  Disallow bar      no-bar

× 3 problems (3 errors, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs results with missing ruleMetadata object 1`] = `
"to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings 1`] = `
"to/file.css
  1:1  ×  Unexpected foo  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings contains non-ASCII characters 1`] = `
"to/file.css
  1:1  ×  简体中文こんにちは안녕하세요  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings for multiple sources 1`] = `
"to/file-a.css
  1:2  ×  Unexpected foo  no-foo

to/file-b.css
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings using the appropriate icon 1`] = `
"to/file.css
  1:1  ‼  Unexpected foo  bar

‼ 1 problem (0 errors, 1 warning)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings with more than 80 characters (no wordbreaks) and \`process.stdout.columns\` equal 90 characters 1`] = `
"to/file.css
  1:1  ×  aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  bar-very-very-very-very-very-long
          aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
          aaaaaaaaaa

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings with more than 80 characters and \`process.stdout.columns\` equal 90 characters 1`] = `
"to/file.css
  1:1  ×  Unexpected very very very very very very very  bar-very-very-very-very-very-long
          very very very very very very long foo

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings with more than 80 non-ASCII characters and \`process.stdout.columns\` equal 90 characters 1`] = `
"to/file.css
  1:1  ×  简体中文こんにちは안녕하세요简体中文こんにち   bar-very-very-very-very-very-long
          は안녕하세요简体中文こんにちは안녕하세요简体
          中文こんにちは안녕하세요简体中文こんにちは안
          녕하세요简体中文こんにちは안녕하세요简体中文
          こんにちは안녕하세요简体中文こんにちは안녕하
          세요简体中文こんにちは안녕하세요

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE outputs warnings without stdout \`TTY\` 1`] = `
"to/file.css
  1:1  ×  Unexpected foo  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using GITHUB_WORKSPACE removes rule name from warning text 1`] = `
"to/file.css
  1:1  ‼  Unexpected foo  rule-name

‼ 1 problem (0 errors, 1 warning)"
`;

exports[`stringFormatter using cwd condenses deprecations and invalid option warnings 1`] = `
"Invalid Option: Unexpected option for baz

Deprecation warnings:
 - Deprecated foo. See: bar"
`;

exports[`stringFormatter using cwd handles empty messages 1`] = `
"path/to/file.css
  1:1  ×     bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs fixable error and warning counts 1`] = `
"path/to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ×  Unexpected bar  no-bar
  1:2  ‼  Unexpected baz  no-baz

× 3 problems (2 errors, 1 warning)
  1 error and 1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using cwd outputs fixable error counts 1`] = `
"path/to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ×  Unexpected bar  no-bar

× 2 problems (2 errors, 0 warnings)
  1 error potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using cwd outputs fixable warning counts 1`] = `
"path/to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)
  1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using cwd outputs fixable warning counts with invalid or missing ruleMetadata 1`] = `
"path/to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar
  1:2  ‼  Unexpected baz  no-baz

× 3 problems (1 error, 2 warnings)
  1 warning potentially fixable with the "--fix" option."
`;

exports[`stringFormatter using cwd outputs parse errors and warnings without rule and severity 1`] = `
"path/to/file.css
  1:1  ×  Cannot parse foo  foo-error
  2:1  ×  Anonymous error
  3:1  ×  Disallow bar      no-bar

× 3 problems (3 errors, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs results with missing ruleMetadata object 1`] = `
"path/to/file.css
  1:2  ×  Unexpected foo  no-foo
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)"
`;

exports[`stringFormatter using cwd outputs warnings 1`] = `
"path/to/file.css
  1:1  ×  Unexpected foo  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs warnings contains non-ASCII characters 1`] = `
"path/to/file.css
  1:1  ×  简体中文こんにちは안녕하세요  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs warnings for multiple sources 1`] = `
"path/to/file-a.css
  1:2  ×  Unexpected foo  no-foo

path/to/file-b.css
  1:2  ‼  Unexpected bar  no-bar

× 2 problems (1 error, 1 warning)"
`;

exports[`stringFormatter using cwd outputs warnings using the appropriate icon 1`] = `
"path/to/file.css
  1:1  ‼  Unexpected foo  bar

‼ 1 problem (0 errors, 1 warning)"
`;

exports[`stringFormatter using cwd outputs warnings with more than 80 characters (no wordbreaks) and \`process.stdout.columns\` equal 90 characters 1`] = `
"path/to/file.css
  1:1  ×  aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  bar-very-very-very-very-very-long
          aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
          aaaaaaaaaa

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs warnings with more than 80 characters and \`process.stdout.columns\` equal 90 characters 1`] = `
"path/to/file.css
  1:1  ×  Unexpected very very very very very very very  bar-very-very-very-very-very-long
          very very very very very very long foo

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs warnings with more than 80 non-ASCII characters and \`process.stdout.columns\` equal 90 characters 1`] = `
"path/to/file.css
  1:1  ×  简体中文こんにちは안녕하세요简体中文こんにち   bar-very-very-very-very-very-long
          は안녕하세요简体中文こんにちは안녕하세요简体
          中文こんにちは안녕하세요简体中文こんにちは안
          녕하세요简体中文こんにちは안녕하세요简体中文
          こんにちは안녕하세요简体中文こんにちは안녕하
          세요简体中文こんにちは안녕하세요

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd outputs warnings without stdout \`TTY\` 1`] = `
"path/to/file.css
  1:1  ×  Unexpected foo  bar

× 1 problem (1 error, 0 warnings)"
`;

exports[`stringFormatter using cwd removes rule name from warning text 1`] = `
"path/to/file.css
  1:1  ‼  Unexpected foo  rule-name

‼ 1 problem (0 errors, 1 warning)"
`;
